master/internal/stream/system.go

```diff
@@ -36,41 +51,100 @@ type SubscriptionSet struct {
 	// Experiments *subscriptionState[*ExperimentMsg, ExperimentSubscriptionSpec]
 }
 
-// subscriptionState contains per-type subscription state
+// subscriptionState contains per-type subscription state.
 type subscriptionState[T stream.Msg, S any] struct {
-	Subscription stream.Subscription[T]
-	FilterMaker FilterMaker[T, S]
-	CollectStartupMsgs CollectStartupMsgsFunc[S]
+	Subscription               stream.Subscription[T]
+	FilterMaker                FilterMaker[T, S]
+	CollectStartupMsgs         CollectStartupMsgsFunc[S]
 	CollectSubscriptionModMsgs CollectSubscriptionModMsgsFunc[S]
 }
 
-type CollectStartupMsgsFunc[S any] func(known string, spec S, ctx context.Context) (
+// CollectStartupMsgsFunc collects messages that were missed prior to startup.
+type CollectStartupMsgsFunc[S any] func(
+	ctx context.Context,
+	user model.User,
+	known string,
+	spec S,
+) (
 	[]*websocket.PreparedMessage, error,
 )
 
-type CollectSubscriptionModMsgsFunc[S any] func(addSpec S, ctx context.Context) (
+// CollectSubscriptionModMsgsFunc collects messages that are missed due to modifying a subscription.
+type CollectSubscriptionModMsgsFunc[S any] func(ctx context.Context, addSpec S) (
 	[]*websocket.PreparedMessage, error,
 )
 
-func NewPublisherSet() PublisherSet {
-	return PublisherSet {
-		Trials: stream.NewPublisher[*TrialMsg](),
+func (ps *PublisherSet) addSocket(socket *websocket.Conn) {
+	ps.socketLock.Lock()
+	defer ps.socketLock.Unlock()
+	ps.activeSockets = append(ps.activeSockets, socket)
+}
+
+// Restart restarts underlying publishers and closes all active websocket connections.
+func (ps *PublisherSet) Restart() (errs []error) {
+	ps.socketLock.Lock()
+	defer ps.socketLock.Unlock()
+	ps.Trials.Restart()
+	// ps.Experiments.Restart()
+
+	// close active websocket connections
+	var remainingSockets []*websocket.Conn
+	for _, socket := range ps.activeSockets {
+		if err := socket.Close(); err != nil {
+			errs = append(errs, err)
+			remainingSockets = append(remainingSockets, socket)
+		}
+	}
+	ps.activeSockets = remainingSockets
+	return errs
```

Comment:
when would this be necessary?  And why have two separate bootem mechanisms?  Can we have just one?

Reply:
Are you talking about the restarting publishers & closing the websockets? 
- If we just close the websockets, streaming clients will know to reconnect but then the server going to hold onto stale Subscriptions in memory
- If we just restart the Publishers, the server will clear out it's known subscriptions but then streaming clients won't know to reconnect since the websocket connections are still open

Or are you talking about keeping the sockets that failed to close?
- It might be worth dropping them and just returning the errors (especially if the failure is due to the websocket connection being closed prior)
   - This is actually something I need to correct, we aren't removing websocket connections from the active socket set unless a Restart is called

Reply:
I'm talking about two separate close mechanisms.

If you close all streamers, why doesn't that close all websockets?

If you close all websockets, why doesn't that close all streamers?

But I didn't even notice the keeping failed-to-close websockets.  Generally, there's no such thing as "failed to close an IO channel".  There can be "the implicit flush associated with the close failed" and there can be "a graceful shutdown was attempted but hit an IO error instead" and there can be "you passed an invalid file descriptor to the close syscall", but generally speaking, the underlying file descriptor SHOULD be closed no matter what.  I haven't read the gorrila code to verify, but it would be a horrible library if that weren't the case.  In general systems code, you normally ignore the return value when you `close()` a read-only file descriptor, and only care about the return value of `close()` when you have potentially unflushed data on a writable file descriptor.

So yeah you should make sure we know what close() means for this particular library and most likely you should ignore the return code... since we already have the assumption that the client is stateful and will reconnect after broken connections.  Or log at debug level, if you like, that might be preferable.

Reply:
Good to know! I went through the gorilla library, it uses the net's Conn & file descriptors under the hood, which makes sure to close the underlying file descriptor, so I'll make sure to log the errors at a debug level & not hold onto the Conn's that had errors during close. 

I also realized that we only drop connections on a restart, which is incorrect, we should be dropping them whenever a websocket is closed.


As far as why closing streamers doesn't close websockets, Streamers/Publishers/Subscriptions don't know the websocket their associated with, in this implementation the only thing that does is the PublisherSet or inside `Websocket()`. Alternatively we could have it setup so that the PublisherSet is listening for the websocket.Conn to close & when it does it proceed to close the underlying Publishers. Is this what you had in mind?

Reply:
Just realized what you meant, I added a `defer socket.Close()` to cleanup the socket when Websocket() returns, so there's no need to keep a set of the active sockets, since by closing the streamers the websocket connection will close